src/fortran/client.F90

```diff
@@ -677,6 +693,102 @@ function set_model_from_file(self, name, model_file, backend, device, batch_size
+  character(kind=c_char, len=len_trim(backend)) :: c_backend
+  character(kind=c_char, len=:), allocatable :: c_tag
+
+  character(kind=c_char, len=:), allocatable, target :: c_inputs(:), c_outputs(:)
+  character(kind=c_char,len=1), target, dimension(1) :: dummy_inputs, dummy_outputs
+
+  integer(c_size_t), dimension(:), allocatable, target :: input_lengths, output_lengths
+  integer(kind=c_size_t) :: name_length, model_file_length, backend_length, tag_length, n_inputs, &
+                            n_outputs
+  integer(kind=c_int)    :: c_batch_size, c_min_batch_size, c_first_gpu, c_num_gpus
+  type(c_ptr)            :: inputs_ptr, input_lengths_ptr, outputs_ptr, output_lengths_ptr
+  type(c_ptr), dimension(:), allocatable :: ptrs_to_inputs, ptrs_to_outputs
+
+  integer :: i
+  integer :: max_length, length
+
+  ! Set default values for the optional inputs
+  c_batch_size = 0
+  if (present(batch_size)) c_batch_size = batch_size
+  c_min_batch_size = 0
+  if (present(min_batch_size)) c_min_batch_size = min_batch_size
+  if (present(tag)) then
+    allocate(character(kind=c_char, len=len_trim(tag)) :: c_tag)
+    c_tag = tag
+    tag_length = len_trim(tag)
+  else
+    allocate(character(kind=c_char, len=1) :: c_tag)
+    c_tag = ''
+    tag_length = 1
+  endif
+
+  ! Cast to c_char kind stringts
+  c_name = trim(name)
+  c_model_file = trim(model_file)
+  c_backend = trim(backend)
+
+  name_length = len_trim(name)
+  model_file_length = len_trim(model_file)
+  backend_length = len_trim(backend)
+
+  ! Convert to C int
+  c_first_gpu = first_gpu
+  c_num_gpus  = num_gpus
+
+  dummy_inputs = ''
+  if (present(inputs)) then
+    call convert_char_array_to_c(inputs, c_inputs, ptrs_to_inputs, inputs_ptr, input_lengths, input_lengths_ptr, &
+                                  n_inputs)
+  else
+    call convert_char_array_to_c(dummy_inputs, c_inputs, ptrs_to_inputs, inputs_ptr, input_lengths, input_lengths_ptr,&
+                                  n_inputs)
+  endif
+
+  dummy_outputs =''
+  if (present(outputs)) then
+    call convert_char_array_to_c(outputs, c_outputs, ptrs_to_outputs, outputs_ptr, output_lengths, output_lengths_ptr,&
+                                  n_outputs)
+  else
+    call convert_char_array_to_c(dummy_outputs, c_outputs, ptrs_to_outputs, outputs_ptr, output_lengths, &
+                                  output_lengths_ptr, n_outputs)
+  endif
+
+  code = set_model_from_file_multigpu_c(self%client_ptr, c_name, name_length, c_model_file, model_file_length, &
+                             c_backend, backend_length, c_first_gpu, c_num_gpus, c_batch_size, c_min_batch_size, &
+                             c_tag, tag_length, inputs_ptr, input_lengths_ptr, n_inputs, outputs_ptr, &
+                             output_lengths_ptr, n_outputs)
+  deallocate(c_inputs)
+  deallocate(input_lengths)
+  deallocate(ptrs_to_inputs)
+  deallocate(c_outputs)
+  deallocate(output_lengths)
+  deallocate(ptrs_to_outputs)
+  deallocate(c_tag)
```

Comment:
Is it safe to unconditionally deallocate memory that was conditionally allocate?

Reply:
Good catch. Fortran standard is to say that it will return an error code but not segfault. That being said, you're right that we should wrap this in `if ALLOCATED(c_outputs) deallocate(c_outputs)`

Reply:
Applies in a couple places and in both this and the non-multi version of the routine